### PR TITLE
forgejo: refactor to ease overriding

### PR DESCRIPTION
**Pull request by nobodyinperson**
## Description of changes
                                                                            
Introduce new optional arguments to `forgejo/generic.nix`:                  
                                                                            
- `rev`: git revision to use, will also be included in the version string.  
- `srcOverride`: for overriding the entire forgejo source.                  
  Note that `overrideAttrs` does not work for this as the frontend npm      
  package must use the same src as the go module but has its source         
  specified in an un-overridable let-binding.                               
- `extraRuntimeInputs`: list of additional packages the forgejo             
  executable should have access to during runtime, e.g. for custom          
  renderers or other extensions (e.g. git-annex).                           

These changes enable overriding the forgejo package to use a specific version or fork and add runtime dependencies without copy-pasting the `generic.nix` from nixpkgs or using a nixpkgs fork entirely, for example:

```nix
  forgejo-aneksajo = (pkgs.callPackage (import
    <nixpkgs/pkgs/by-name/fo/forgejo/generic.nix> rec {
      version = "8.0.2-git-annex0";
      rev = "b62d89ce67";
      hash = "sha256-kjvTPT5Iw2rm3ksP+9mm1tQOTLtr8T2Zikw5dl6aFWI=";
      srcOverride = pkgs.fetchFromGitea {
        domain = "codeberg.org";
        owner = "nobodyinperson";
        repo = "forgejo-aneksajo";
        inherit rev hash;
      };
      npmDepsHash = "sha256-dIf1Sh0jrQPM2IY+VtTtrEioBM1r43mQEI0fG4DcGTc=";
      vendorHash = "sha256-4l4kscwesW/cR8mZjE3G9HcVm0d1ukxbtBY6RXYRi8k=";
      lts = false;
      extraRuntimeInputs = [ pkgs.git-annex ]; # for annex support
    }) { }).overrideAttrs (oldAttrs: { doCheck = false; });
```

The `forgejo` and `forgejo-lts` packages still evaluate to the same, so no rebuild should be necessary. 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [x] [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - [x] `nix-build -A nixosTests.forgejo`
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
    - [x] `nix-build -A pkgs.forgejo.passthru.tests`
    - [x] `nix-build -A pkgs.forgejo-lts.passthru.tests`
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/fo/forgejo/generic.nix [hunk 1]
```diff
@@ -1,9 +1,12 @@
 {
   lts ? false,
   version,
+  rev ? null,
+  srcOverride ? null,
```
Comment (Atemu) [inline pkgs/by-name/fo/forgejo/generic.nix:5]: To achieve something similar with different intentions, I used this pattern in ffmpeg:

https://github.com/NixOS/nixpkgs/blob/5a24308c384447887fe59453dd23b5cb09079593/pkgs/development/libraries/ffmpeg/generic.nix#L3-L11

I find it quite a bit more idiomatic and simple.

Comment (nobodyinperson) [reply pkgs/by-name/fo/forgejo/generic.nix:5]: Oh cool I didn't know that one can reuse attributes directly in the function args attrset, I'll rewrite this like you did. 👍

Comment (nobodyinperson) [reply pkgs/by-name/fo/forgejo/generic.nix:5]: Ah, problem is that `fetchFromGitea` is not available up there, your `fetchgit` is a builtin.

**Review by adamcstephens — CHANGES_REQUESTED** [review]
Except for the npmdepsHash, you can use standard overrideAttrs for this without any changes. Extra dependencies should be added to the service/environment or you could rewrap forgejo, which has the benefit of not forcing an entire package rebuild when changed on your end.

```
forgejo = inputs'.nixpkgs-unstable.legacyPackages.forgejo.overrideAttrs (old: {
  name = "someother";
  src = pkgs.fetchFromGitea {
    domain = "codeberg.org";
    owner = "nobodyinperson";
    repo = "forgejo-aneksajo";
    ...
  };
});
```

nixfmt changes should not be included regardless.

**Review by emilylange — CHANGES_REQUESTED** [review]
Hard no from me.

I am not willing to help maintain niche external use-cases that built on top of an internal abstraction.
I've been there before, it's painful.
There are no stable guarantees for that internal generic.nix file and I would like to keep it that way.

The actual Forgejo derivation isn't all that long, so I'd recommend copying it to your local config instead, if you that's easier for you.

I can assure you, what you are trying to accomplish is already possible with `overrideAttrs`, it's just very hacky.
Most notably, you could replace that `cp -R ./{templates,options} ${frontend}/public $data` line in `postInstall` to point to your own frontend derivation output.

**Comment by nobodyinperson** [comment]
> nixfmt changes should not be included regardless.

Good to know. The contribution guidelines said to run `nixfmt`. 🤷

> I can assure you, what you are trying to accomplish is already possible with overrideAttrs, it's just very hacky.

Yep, hacky it is.

> Extra dependencies should be added to the service/environment

Good point, I didn't think of that 👍

**Comment by nobodyinperson** [comment]
Just for reference, I indeed managed to do what I want with just `overrideAttrs`, and I find the hackiness level is reasonable:

https://gitlab.com/nobodyinperson/nixconfig/-/blob/2739a2e8df08a007d3003f3fc7ef5b04de57e88e/forgejo.nix#L5-L60

Anyway, thanks for your time reviewing this PR and sorry to have bothered you 🙏

**Pull request closed without merging**